### DISCUSSION TITLE
Bulk operations

### DISCUSSION THREAD

**Discussion by jmikrut** [payloadcms/payload#1382] [Roadmap]
We need to create bulk create / update / delete operations and extend admin UI to support bulk actions.

Relevant issues:
- https://github.com/payloadcms/payload/issues/1365

Marked as `planned`.

**Comment by manuelebagnolini** [comment]
I think that the bulk operations should be extended to the localizations in order to allow the massive insertion/updating of all the languages ​​of the document in a single request.

**Comment by jmikrut** [comment]
Update here - this is very important and is high priority on our roadmap.

Work involved will be broken down into the following areas:

1. Bulk update from `List` view - add checkboxes to select multiple docs, and then open a modal to edit individual documents all at once
2. Bulk delete from the `List` view - again, checkboxes to select multiple docs to delete
3. Figure out how to handle this on the backend. New bulk operations? Parallel REST requests from the admin UI? How often would we use this? If we wanted to batch edit 5 documents, and 4 worked, but one failed, should all 5 fail? Or should the 4 still update and have the 1 fail only?

**Comment by reiv** [comment]
Will these bulk operations be available in the local API as well?

**Reply by jmikrut** [reply]
Yes!

**Comment by AlessioGr** [comment]
What do you think about making the list view work kinda like a spreadsheet, so you can edit all fields all at once? Would be such a huge time-saver

**Reply by jmikrut** [reply]
This is pretty much what we're envisioning, give or take. We'll be working on UX and opening an RFC for thoughts shortly!

**Reply by AlessioGr** [reply]
> This is pretty much what we're envisioning, give or take. We'll be working on UX and opening an RFC for thoughts shortly!

Awesome! You think you'd wanna take it to a level, where it could replace tools like [Smartsheet](https://www.smartsheet.com/)?
For example, I currently have this in Smartsheet which is supposed to track Freelancer progress on content:
![image](https://user-images.githubusercontent.com/70709113/204654102-1f342702-79db-421f-9c4b-b08ce0a0453f.png)

Having something like that in Payload would be insanely helpful. As I could then just use it just like data, I could easily use that data to build frontends for freelancers & dashboards for myself, or reference to it from other places from payload.

Or would you like to keep it more simple in payload? What do you think?

**Reply by jmikrut** [reply]
It's a good question. I think there are really two things that need to be combined here to make a solid set of features:
1. changing a value across  _**multiple documents at once**_
2. allowing for easier inline editing within the List view 

This spreadsheet-style view is a solution for 2, but not necessarily for 1. Know what I mean?

**Reply by AlessioGr** [reply]
> It's a good question. I think there are really two things that need to be combined here to make a solid set of features:
> 
> 1. changing a value across  _**multiple documents at once**_
> 2. allowing for easier inline editing within the List view
> 
> This spreadsheet-style view is a solution for 2, but not necessarily for 1. Know what I mean?

Hmm what would a solution for 1. look like? I was imagining the same kind of spreadsheet, where you are able to select multiple cells and change their values at once.

**Reply by jmikrut** [reply]
Generally a solution for 1 would be to select the "rows" from the table that you want to edit via checkbox or similar, and then a little "drawer" slides out, where you could choose which fields you want to edit. Then the drawer would have a way to edit the value for only the fields that you select to edit. You'd enter one value and then that one value would be applied to all selected documents.

I think the difference would be that in 1, you are setting the same value across many documents, but in 2, you are free to type in a different value for each selected document.

**Reply by AlessioGr** [reply]
> Generally a solution for 1 would be to select the "rows" from the table that you want to edit via checkbox or similar, and then a little "drawer" slides out, where you could choose which fields you want to edit. Then the drawer would have a way to edit the value for only the fields that you select to edit. You'd enter one value and then that one value would be applied to all selected documents.
> 
> I think the difference would be that in 1, you are setting the same value across many documents, but in 2, you are free to type in a different value for each selected document.

Ahh I see! I think 2 would actually have the best UX for stuff like text or number fields, if you want to edit multiple at once. But then 1. would be great for more complicated fields, e.g. Upload fields or Relationships.

I think you can modify a lot of fields with 2. if done well, as long as it's text content, e.g. Dropdowns. And for more complicated fields: what if, once the complicated fields are selected (like a spreadsheet), a little button pops-up which then opens the edit drawer? So, instead of using the drawer to choose the fields, you choose the fields naturally by selecting them in the spreadsheet. Would also save more time, I think that's a nice way to combine it.

If that's done well, this could really be HUGE. E.g. a bigger company I'm working for part-time is using Google Sheets for things which should actually be data in a database. But no, they use sheets for data (account managers access and update it) which is a pain to parse and use somewhere else if you need its data.
Payload having that functionality would be a killer feature

Now replacing a generalist app like Google Sheets would probably be too much (for now?), but any sheet which is actually defined data (so anything which has a defined set of columns) would be great for payload to handle

**Comment by jcouturier** [comment]
Would this new feature apply to media uploads as well, like being able to bulk upload media?

**Reply by DanRibbens** [reply]
The initial work I have for this is not covering `create` just yet. I will see if I can squeeze this in after what I am working on.

So far I'm working through:

- API support
	- [x] Delete operations
	- [x] Edit operations
- UI
	- [x] List view selection of many documents
	- [x] Allow for queries / select all
	- [x] Delete selected
	- [x] Edit selected, choosing the fields you want to edit (WIP)
- With versions enabled:
	- [x] Publish
	- [x] Unpublish

**Reply by DanRibbens** [reply]
Oops, I didn't answer on your question about uploads. I need to test and think about what is happening in the edit workflow for upload fields.

Creating many from a single API request feels like a whole new feature altogether and I'm not sure how it work from a UI standpoint either. I'm going to keep these thoughts out until later, same for submitting multiple locals. 

There is another roadmap item on uploading multiple files at once: https://github.com/payloadcms/payload/discussions/1149

**Comment by wwgleah** [comment]
What's the ETA on the bulk operations feature?

**Reply by DanRibbens** [reply]
This is very close to being ready. I will open a pull request tomorrow for review.

**Reply by DanRibbens** [reply]
I found a few more things that needed to be addressed these past few days&mdash;happy to say this is now code complete and now ready for review.

**Comment by DanRibbens** [comment]
Bulk operations are available now in [v1.6.24](https://github.com/payloadcms/payload/releases/tag/v1.6.24).

I hope you all enjoy!

**Comment by wwgleah** [comment]
Incredible!!!! Thank you

On Thu, Mar 23, 2023 at 8:28 PM Dan Ribbens ***@***.***>
wrote:

> Bulk operations are available now in v1.6.24
> <https://github.com/payloadcms/payload/releases/tag/v1.6.24>.
>
> I hope you all enjoy!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/payloadcms/payload/discussions/1382#discussioncomment-5412628>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6P2FPWNG437MP3OAGKSKA3W5T2DJANCNFSM6AAAAAAR6N3PNE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
*Leah Lykins*
She/Her | Co-Founder | WhereWeGo <https://www.wherewego.org> | c. 504 505
9422

WhereWeGo is an EdTech and design thinking firm built by former teachers.
Its flagship platform, the Training Navigator
<https://explore.wherewego.org>, was created by a community of young
adults, counselors, nonprofit leaders, and career education providers
across New Orleans.

*The future of labor is the future of America. - John Lewis.*

**Comment by JeremieLeblanc** [comment]
Are there still plans to expand this functionality to include create?
Seems like you should be able to pass an array of documents to the create endpoint.

I understand the error handling is the pain point in this scenario.
Do you treat it as a transaction where it's either all or nothing, or do you allow partial success...
Either way I think it would be a welcomed addition.

**Comment by dan-scrumsdotcom** [comment]
Would love a bulk create.

**Comment by Saissaken** [comment]
@DanRibbens is this still supported? is it planned for the new payload?

**Comment by thomas-coldwell** [comment]
This is something that I would really love to see for relationship fields (especially with the drawer appearance). Right now your options for adding documents to a relationship field are:
1. Add each one manually by clicking on the field selecting one and then rinse repeat... not really scalable for 1000's of docs
2. Write some custom tooling / view to go and do this with the local API

As you can already select documents and do this with a filter + search in the collection view there it feels like the same UI and functionality is then missing when trying to add relationships from this drawer view

**Reply by thomas-coldwell** [reply]
Linking this related discussion I found asking for this too https://github.com/payloadcms/payload/discussions/12787
